### DISCUSSION TITLE
Conditionally Apply AsParallel

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/runtime#55766] [Q&A]
I have this code:
```
        private static IEnumerable<T> ApplyParallel<T>(IEnumerable<T> Source, bool Parallel) {
            var ret = Source;

            if (Parallel) {
                ret = ret.AsParallel()
                    .WithExecutionMode(ParallelExecutionMode.ForceParallelism)
                    .WithMergeOptions(ParallelMergeOptions.FullyBuffered)
                    ;
            }

            return ret;
        }
```

But when I use it like so:
```
(from x in ApplyParallel(Source, true) ...)
```
I don't get parallelism.  But if I use it like so:
```
(from x in Source.AsParallel()
                    .WithExecutionMode(ParallelExecutionMode.ForceParallelism)
                    .WithMergeOptions(ParallelMergeOptions.FullyBuffered)
...)
```
I do. 

Is there a way to make my method work as intended?

**Comment by Joe4evr** [comment]
Your method returns plain `IEnumerable<T>` which makes your query bind to the serialized `Enumerable` methods, but if you want parallelism it needs to be `ParallelQuery<T>` so that it'll bind to the `ParallelEnumerable` methods.

So no, you cannot conditionally make a query parallel this way.

**Reply by svick** [reply]
You could use `.WithDegreeOfParallelism(1)` to force a `ParallelQuery<T>` to not execute in parallel. I think that's the best that you can do here.

**Reply by TonyValenti** [reply]
Thanks @svick !  That's what I'll do.
